### PR TITLE
Support course copy for Canvas pages

### PR DESCRIPTION
**Pull request by marcospri**
Course copy is a feature that most LMS have that allows to... well, copy a course. This is useful to prepare courses for new terms.

In general the content of the course gets copied alongside the assignments. Our assignments that point to a document in the LMS, lets day "page id 1", will still point to that document, the old course one, in the new course. Without any intervention that will make the assignment load to fail.

To fix this issue we try to find a equivalent page in the new course, like we do files. We'll base this on the title. 

Once we found the new page, we store that information in the DB so consequent launches don't need to find that information again.

## Testing

- Launch https://hypothesis.instructure.com/courses/606/assignments/5710 in `main`

You'll get an error 

- Switch to this branch, `canvas-pages-find-page`, launch again.

The assignment should work, you'll get something along the lines of

```Via URL for page, found page in the new course. Document: canvas://page/course/125/page_id/1479, course: 606, new page id: 1481```

- Launch again, it should work again but it will now use the information from the previous launch:

```
Via URL for page already mapped for course copy. Document: canvas://page/course/125/page_id/1479, course: 606, mapped page_id: 1481
```

- Launch the original assignment, no course copy, it should work as before:

https://hypothesis.instructure.com/courses/125/assignments/5632

logging something like:

`Via URL for page in the same course. canvas://page/course/125/page_id/1479`

### CODE REVIEW

File: lms/views/api/canvas/pages.py [hunk 1]
```diff
@@ -43,11 +51,72 @@ def list_pages(self):
...
         assignment = self.request.find_service(name="assignment").get_assignment(
-            application_instance.tool_consumer_instance_guid,
+            self.request.lti_user.application_instance.tool_consumer_instance_guid,
             self.request.lti_user.lti.assignment_id,
         )
+        document_url = assignment.document_url
+        document_course_id, document_page_id = self._parse_document_url(document_url)
+
+        effective_page_id = None
+        if current_course_id == document_course_id:
```
Comment (marcospri) [inline lms/views/api/canvas/pages.py:69]: If the course in the document_url is the same as in the launch, just use the info from the document_url. No course copy here.

File: lms/views/api/canvas/pages.py [hunk 2]
```diff
@@ -43,11 +51,72 @@ def list_pages(self):
...
+            # Not in a course copy scenario, user the IDs from the document_url
+            effective_page_id = document_page_id
+            LOG.debug("Via URL for page in the same course. %s", document_url)
+
+        mapped_page_id = current_course.get_mapped_page_id(document_page_id)
+        if not effective_page_id and mapped_page_id != document_page_id:
```
Comment (marcospri) [inline lms/views/api/canvas/pages.py:75]: If we are in course copy but we stored an ID previously, use that one.

File: lms/views/api/canvas/pages.py [hunk 3]
```diff
@@ -43,11 +51,72 @@ def list_pages(self):
...
+            effective_page_id = mapped_page_id
+            LOG.debug(
+                "Via URL for page already mapped for course copy. Document: %s, course: %s, mapped page_id: %s",
+                document_url,
+                current_course_id,
+                mapped_page_id,
+            )
+
+        if not effective_page_id:
```
Comment (marcospri) [inline lms/views/api/canvas/pages.py:84]: Course copy, first launch (no stored ID)

File: lms/views/api/canvas/pages.py [hunk 4]
```diff
@@ -43,11 +51,72 @@ def list_pages(self):
...
+            found_page = course_copy_plugin.find_matching_page_in_course(
+                document_page_id, current_course_id
+            )
+            if not found_page:
```
Comment (marcospri) [inline lms/views/api/canvas/pages.py:88]: No luck

File: lms/views/api/canvas/pages.py [hunk 5]
```diff
@@ -60,7 +129,8 @@ def via_url(self):
                 self.request.route_url(
                     "canvas_api.pages.proxy",
                     _query={
-                        "document_url": assignment.document_url,
+                        "course_id": current_course_id,
```
Comment (marcospri) [inline lms/views/api/canvas/pages.py:132]: The proxy endpoints changes, from being based on the document_id to be explicitly told which course and page ids to use.

**Review by acelaya — APPROVED** [review]
LGTM.

The only thing is that I do not get an error `main` when launching the assignment. It does work.

I did a couple of attempts, so it may be related with that.

**Comment by marcospri** [comment]
> The only thing is that I do not get an error main when launching the assignment. It does work.

My bad, the instructors are missing steps.

Adding them here:

- Launching https://hypothesis.instructure.com/courses/606/assignments/5710 in main success as a teacher.

The teacher has access to the original page in the original course. They are annotating that document instead of the one in the new course. This *works* but it's broken in surprising ways, eg changes on the page in this course won't have any effect.


- Launching https://hypothesis.instructure.com/courses/606/assignments/5710 as `eng+coursecopystudent` fails in `main` as that student is not enrolled in the original course.

**Pull request merged**